homeassistant/components/sensor/demo.py

```diff
@@ -4,16 +4,68 @@
 For more details about this platform, please refer to the documentation
 https://home-assistant.io/components/demo/
 """
+import datetime as datetime
+import math
+import random
+import logging
+
+import voluptuous as vol
 from homeassistant.const import ATTR_BATTERY_LEVEL, TEMP_CELSIUS
 from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_NAME
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = datetime.timedelta(seconds=1)
+ICON = 'mdi:chart-line'
+
+CONF_UNIT = 'unit'
+CONF_AMP = 'amplitude'
+CONF_MEAN = 'mean'
+CONF_PERIOD = 'period'
+CONF_PHASE = 'phase'
+CONF_FWHM = 'spread'
+CONF_SEED = 'seed'
+
+DEFAULT_NAME = 'simulated'
+DEFAULT_UNIT = 'value'
+DEFAULT_AMP = 1
+DEFAULT_MEAN = 0
+DEFAULT_PERIOD = 60
+DEFAULT_PHASE = 0
+DEFAULT_FWHM = 0
+DEFAULT_SEED = 999
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_UNIT, default=DEFAULT_UNIT): cv.string,
+    vol.Optional(CONF_AMP, default=DEFAULT_AMP): vol.Coerce(float),
+    vol.Optional(CONF_MEAN, default=DEFAULT_MEAN): vol.Coerce(float),
+    vol.Optional(CONF_PERIOD, default=DEFAULT_PERIOD): cv.positive_int,
+    vol.Optional(CONF_PHASE, default=DEFAULT_PHASE): vol.Coerce(float),
+    vol.Optional(CONF_FWHM, default=DEFAULT_FWHM): vol.Coerce(float),
+    vol.Optional(CONF_SEED, default=DEFAULT_SEED): cv.positive_int,
+})
 
 
 # pylint: disable=unused-argument
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the Demo sensors."""
+    name = config.get(CONF_NAME)
+    unit = config.get(CONF_UNIT)
+    amp = config.get(CONF_AMP)
+    mean = config.get(CONF_MEAN)
+    period = config.get(CONF_PERIOD)
+    phase = config.get(CONF_PHASE)
+    fwhm = config.get(CONF_FWHM)
+    seed = config.get(CONF_SEED)
+    random.seed(seed)
```

Comment:
I think having this seed value being used for the global random instance is misleading. What if some code somewhere also accesses the global random instance? Then all of our supposedly reproducible (due to the seed config entry) random values become irreproducible again. See https://stackoverflow.com/questions/12368996/what-is-the-scope-of-a-random-seed-in-python

Also: Is there a use case for seeds anyway? I mean I can't think of any...

Reply:
The use case is when a developer (e.g. of a filter) wants to ensure that the data produced by the component is identical on each run. However that same developer could also just edit the file if they wished so happy to remove if preferred

Reply:
I understand the use case, but I think *if* we add a seed config value, we should make sure that it actually works correctly and produces the same output on each run. And I think that's only the case once the sensor produces the same output **independent** of any other code being run on the hass instance.

Additionally, I don't think it would be much work to use a local PRNG instead of the global `random` one. Just create do *something like*:

```python
prng = random.random(config.get(CONF_SEED))
# ...
SimulatedSensor(name, unit, amp, mean, period, phase, fwhm, prng)
```

Reply:
OK followed https://stackoverflow.com/questions/12368996/what-is-the-scope-of-a-random-seed-in-python